### PR TITLE
samples: bluetooth: mesh: sensor sample features

### PR DESCRIPTION
**Pull request by HaavardRei**
Updates the sensor sample with a range limit setting property
for the chip temperature sensor.

Adds a descriptor to the chip temperature sensor.

The client sample sends:
A get sensor range setting message on button 1 press.
A set sensor range setting message on button 2 press.
A get sensor descriptor message on button 3 press.

Adds functionality for persistently storing the range limit setting.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

### CODE REVIEW

File: samples/bluetooth/mesh/sensor_server/src/model_handler.c [hunk 1]
```diff
@@ -66,20 +80,100 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
...
+	.set = chip_temp_range_set,
+} };
+
+static int chip_temp_range_settings_restore(const char *name, size_t len, settings_read_cb read_cb,
+					    void *cb_arg)
+{
+	const char *next;
+	int rc;
+
+	if (settings_name_steq(name, "range", &next) && !next) {
```
Comment (PavelVPV) [inline samples/bluetooth/mesh/sensor_server/src/model_handler.c:140]: Nit-picking. I think common style is:
```
if (/* error condition */) {
    return -ENOENT;
}

...

return rc;
```

**Review by PavelVPV — COMMENTED** [review]
Looks good! Just few minor comments from my side.

**Review by Andrewpini — APPROVED** [review]
I have already reviewed this in @HaavardRei's fork. From my side it looks good but need a couple of fresh eyes on it as well.

File: samples/bluetooth/mesh/sensor_server/src/model_handler.c (hunk 1 above)
Comment (HaavardRei) [reply samples/bluetooth/mesh/sensor_server/src/model_handler.c]: I've tried to fix it now!

File: samples/bluetooth/mesh/sensor_server/src/model_handler.c [hunk 2]
```diff
@@ -66,9 +82,91 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
...
+
+		rc = read_cb(cb_arg, &temp_range, sizeof(temp_range));
+		if (rc < 0) {
+			return rc;
+		} else {
+			printk("Restored temperature range setting\n");
+			return 0;
+		}
+	}
+}
```
Comment (Andrewpini) [inline samples/bluetooth/mesh/sensor_server/src/model_handler.c:157]:
```suggestion
		rc = read_cb(cb_arg, &temp_range, sizeof(temp_range));
		if (rc < 0) {
			return rc;
		} 
		
		printk("Restored temperature range setting\n");
		return 0;
	}
}
```

File: samples/bluetooth/mesh/sensor_server/src/model_handler.c [hunk 3]
```diff
@@ -66,9 +82,91 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
...
+		}
+
+		rc = read_cb(cb_arg, &temp_range, sizeof(temp_range));
+		if (rc < 0) {
+			return rc;
+		} else {
+			printk("Restored temperature range setting\n");
+			return 0;
+		}
+	}
```
Comment (Andrewpini) [inline samples/bluetooth/mesh/sensor_server/src/model_handler.c:156]:
```suggestion
	if (!(settings_name_steq(name, "range", &next) && !next)) {
		return -ENOENT;
	}
	
	if (len != sizeof(temp_range)) {
		return -EINVAL;
	}

	rc = read_cb(cb_arg, &temp_range, sizeof(temp_range));
	if (rc < 0) {
		return rc;
	}
	
	printk("Restored temperature range setting\n");
	return 0;
```

File: samples/bluetooth/mesh/sensor_client/README.rst [hunk 4]
```diff
@@ -80,6 +81,17 @@ Buttons:
...
+
+Button 1:
+   Sends a setting get message.
+
+Button 2:
+   Sends a setting set message, switching between the ranges specified in the :c:var:`temp_ranges` variable.
+
+Button 3:
+   Sends a descriptor get message, requesting info about the :c:var:`bt_mesh_sensor_present_dev_op_temp` sensor.
+
```
Comment (mia-ko) [inline samples/bluetooth/mesh/sensor_client/README.rst:94]:
```suggestion
Once the provisioning procedure has completed, the buttons will have the following functionality:

Button 1:
   Sends a get message for a setting.

Button 2:
   Sends a set message for a setting, switching between the ranges specified in the :c:var:`temp_ranges` variable.

Button 3:
   Sends a get message for a descriptor, requesting information about the :c:var:`bt_mesh_sensor_present_dev_op_temp` sensor.

```

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Review by mia-ko — COMMENTED** [review]
Also, we might want to add a changelog entry for the changes in these two samples.

**Comment by HaavardRei** [comment]
@mia-ko @Andrewpini I added the changes that you suggested (and tested the sample). How would I go about adding a changelog entry?

**Review by PavelVPV — APPROVED** [review]

**Comment by HaavardRei** [comment]
Last force push was to fix a compliance warning.

**Comment by mia-ko** [comment]
> @mia-ko @Andrewpini I added the changes that you suggested (and tested the sample). How would I go about adding a changelog entry?

For changelog, add an entry with change description to file ncs\nrf\doc\nrf\releases\release-notes-changelog.rst. There is an own section for Bluetooth mesh samples.
Reference the samples changed (:ref:`bluetooth_mesh_sensor_client` and :ref:`bluetooth_mesh_sensor_server`).

**Review by mia-ko — APPROVED** [review]

File: samples/bluetooth/mesh/sensor_client/src/model_handler.c [hunk 5]
```diff
@@ -58,11 +82,25 @@ static struct k_work_delayable get_data_work;
...
+		return;
+	}
+
+	static uint32_t sensor_idx;
 	int err;
 
 	/* Only one message can be published at a time. Swap sensor after each timeout. */
-	if (req_temp_range) {
+	switch ((sensor_idx++ % 3)) {
+	case (0): {
```
Comment (ludvigsj) [inline samples/bluetooth/mesh/sensor_client/src/model_handler.c:95]: Why the parentheses?

**Review by ludvigsj — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mesh/sensor_server/README.rst [hunk 6]
```diff
@@ -8,8 +8,8 @@ Bluetooth: Mesh sensor
    :depth: 2
 
 The Bluetooth® mesh sensor sample demonstrates how to set up a basic mesh Sensor Server model application that provides sensor data to one :ref:`bt_mesh_sensor_cli_readme` model.
-Four different sensor types are used to showcase different ways for the server to publish data.
-In addition, the sample demonstrates usage of both :ref:`single-channel sensor types and sensor series types <bt_mesh_sensor_types_channels>`.
+Five different sensor types are used to showcase different ways for the server to publish data.
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply samples/bluetooth/mesh/sensor_server/README.rst:11]: Sure, changed it to four. The sensor types, including the setting, is elaborated on in the overview section.

File: samples/bluetooth/mesh/sensor_client/src/model_handler.c (hunk 5 above)
Comment (HaavardRei) [reply samples/bluetooth/mesh/sensor_client/src/model_handler.c:95]: Operator precedence is scary! Removed them now.

File: samples/bluetooth/mesh/sensor_client/src/model_handler.c [hunk 7]
```diff
@@ -71,21 +109,68 @@ static void get_data(struct k_work *work)
...
+static void button_handler_cb(uint32_t pressed, uint32_t changed)
+{
+	if (!bt_mesh_is_provisioned()) {
+		return;
+	}
+
+	static uint32_t temp_idx;
+	int err;
+
+	if ((pressed & BIT(0))) {
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply samples/bluetooth/mesh/sensor_client/src/model_handler.c:144]: Fixed!

File: samples/bluetooth/mesh/sensor_server/src/model_handler.c [hunk 8]
```diff
@@ -54,6 +63,13 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
 		printk("Error getting temperature sensor data (%d)\n", err);
 	}
 
+	if (!bt_mesh_sensor_value_in_column(rsp, &temp_range)) {
+		if (rsp->val1 <= temp_range.start.val1) {
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply samples/bluetooth/mesh/sensor_server/src/model_handler.c]: Should be good now!

**Review by ludvigsj — APPROVED** [review]

**Pull request merged**